exec/csnex/Bytecode.cs

```diff
@@ -0,0 +1,140 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace csnex
+{
+    public class Bytecode
+    {
+        private const int BYTECODE_VERSION = 3;
+        private byte[] obj;
+        public string source_path;
+        public List<String> strtable;
+        public byte[] code;
+
+        public List<FunctionInfo> functions;
+
+        public Bytecode()
+        {
+        }
+
+        public static UInt32 Get_VInt(byte[] pobj, ref uint i)
+        {
+            uint r = 0;
+            while (i < pobj.Length) {
+                byte x = pobj[i];
+                i++;
+                if ((r & ~(uint.MaxValue >> 7)) != 0) {
+                    throw new BytecodeException(string.Format("Integer value exceeds maximum ({0})", int.MaxValue));
+                }
+                r = (r << 7) | (uint)(x & 0x7f);
+                if ((x & 0x80) == 0) {
+                    break;
+                }
+            }
+            return r;
+        }
+
+        public List<String> GetStrTable(byte[] obj, uint size, ref uint i)
+        {
+            List<String> r = new List<string>();
+
+            while (i < size) {
+                uint len = Get_VInt(obj, ref i);
+                string ts = new string(System.Text.Encoding.UTF8.GetChars(obj, (int)i, (int)len));
+                r.Add(ts);
+                i += len;
+            }
+            return r;
+        }
+
+        public void LoadBytecode(string a_source_path, byte[] bytes)
+        {
+            source_path = a_source_path;
+            obj = bytes;
+            uint i = 0;
+
+            if (i + 4 > obj.Length) {
+                throw new BytecodeException("unexpected end of bytecode");
+            }
+            string sig = new string(System.Text.Encoding.ASCII.GetChars(obj, (int)i, 4));
+            if (sig != "Ne\0n") {
+                throw new BytecodeException("bytecode signature missing");
+            }
+            i += 4;
+
+            uint version = Get_VInt(obj, ref i);
+
+            if (version != BYTECODE_VERSION) {
+                throw new BytecodeException("bytecode version mismatch");
+            }
+
+            if (i + 32 > obj.Length) {
+                throw new BytecodeException("unexpected end of bytecode");
+            }
+
+            byte[] source_hash = new byte[32];
+            Array.Copy(obj, i, source_hash, 0, 32);
+            i += 32;
+
+            /* Globals */
+            Get_VInt(obj, ref i);
+
+            /* String Table */
+            uint strtablesize = Get_VInt(obj, ref i);
+            strtable = GetStrTable(obj, strtablesize + i, ref i);
+
+            /* Types */
+            Get_VInt(obj, ref i);
```

Comment:
It would be helpful, at least in development, to have an assertion here that the integer read from the bytecode is in fact 0. If it's not, you can detect an unimplemented bytecode section immediately rather than some other failure later.

Reply:
Added to all of the `Get_Vint()`'s.